opt: fix FK cascades to child tables with partial indexes

mgartner opened the pull request:
Previously, the optimizer was not synthesizing partial index DEL columns
for FK cascading updates and deletes. As a result, a cascading `UPDATE`
could corrupt a child table's partial index, ultimately resulting in
incorrect query results. A cascading `DELETE` would not corrupt
partial indexes, but unnecessary `DEL` operations would be issued on
the partial index.

The optbuilder has been refactored so that these columns are correctly
projected. There are now three functions for projecting PUT columns, DEL
columns, and both PUT and DEL columns, each ensuring that the input
scopes are non-nil. These three functions are called from principal
functions in the optbuilder where CHECK constraint columns are also
projected, like `mutationBuilder.buildUpdate`. In theory this should
make it harder in the future to omit these necessary projections.

Additionally, the execution engine was unable to handle extraneous
columns that can be added as input to FK cascading updates. These
extraneous columns would be incorrectly interpreted as synthesized
partial index columns. This commit works around this issue by slicing
the source values with an upper bound in `updateNode.processSourceRow`.
The longer term fix is to not produce these columns (see issue #57097).

Fixes #57085
Fixes #57084

Release justification: This is a critical bug fix to a new feature,
partial indexes.

Release note (bug fix): A bug has been fixed that caused errors or
corrupted partial indexes of child tables in foreign key relationships
with cascading `UPDATE`s and `DELETE`s. The corrupt partial indexes
could result in incorrect query results. Any partial indexes on child
tables of foreign key relationships with `ON DELETE CASCADE` or `ON
UPDATE CASCADE` actions may be corrupt and should be dropped and
re-created. This bug was introduce in version 20.2.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57100)
<!-- Reviewable:end -->

mgartner commented:
Ooops. Forgot to include some optbuilder tests. I'll add those.

mgartner commented:
There might also be problems with an `ON CASCADE DELETE` but I'm having trouble producing a test case exhibiting a bug.

rytaft reviewed (commented):
[nit] in the commit/PR message, this sentence is garbled: "... columns have been synthesized for from the optimizer to the execution engine ..."

> There might also be problems with an ON CASCADE DELETE but I'm having trouble producing a test case exhibiting a bug.

Maybe you can use `SHOW TRACE`?

> Reviewed 14 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-:-MN-1lw01reFtQ09yWtn:b7rysxe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/update.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MN-2_hnCMDTqeXNDcc_:-MN-2_hnCMDTqeXNDcca:b-i1lff3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/update.go#L59)):*
> ```Go
> 
> 	checkOrds        checkSet
> 	partialIndexOrds partialIndexSet
> ```

Is this getting used other than for its length?
___
*[pkg/sql/logictest/testdata/logic_test/partial_index, line 1530 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MN-1FmX9WQ9-vr_WAGX:-MN-1FmY444wbSl4TtGi:b921m9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/logictest/testdata/logic_test/partial_index#L1530)):*
> ```
> INSERT INTO t57085c_groups VALUES (1), (2);
> INSERT INTO t57085c_users VALUES (10, 1, 100), (20, 2, 200);
> UPDATE t57085c_groups SET id = 3 WHERE id = 1;
> ```

should there be a similar test to ensure that rows are deleted from the partial index if group_id is updated from 3->4?
___
*[pkg/sql/opt/exec/factory.opt, line 416 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MMzyRqe7LBJ7cIOng1T:-MMzyRqe7LBJ7cIOng1U:bl9aqu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/exec/factory.opt#L416)):*
> ```
> # transaction (if appropriate, i.e. if it is in an implicit transaction).
> # This is false if there are multiple mutations in a statement, or the output
> # of the mutation is processed through side-effecting expressions.
> ```

[nit] update comment to mention partial indexes
___
*[pkg/sql/opt/exec/execbuilder/mutation.go, line 343 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MMzybb30FdRXmvcT-Mx:-MMzybb30FdRXmvcT-My:b-mr3lin) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/exec/execbuilder/mutation.go#L343)):*
> ```Go
> 		returnColOrds,
> 		checkOrds,
> 		partialIndexPutOrds,
> ```

why do we only need `Put` here and not `Del`?
___
*[pkg/sql/opt/optbuilder/fk_cascade.go, line 637 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MMzz1fz77_7XZd5whcQ:-MMzz1fz77_7XZd5whcR:b9qkes7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/optbuilder/fk_cascade.go#L637)):*
> ```Go
> 		)
> 
> 		// Set the fetchScope to the scope of the fetch expression.
> ```

what's a fetch expression?
___
*[pkg/sql/opt/optbuilder/insert.go, line 908 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MMzzvrJ8LrH4k2o40bI:-MMzzvrJ8LrH4k2o40bJ:b-lg8hg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/optbuilder/insert.go#L908)):*
> ```Go
> 	//       reason other than as "fetch columns". See buildScan comment.
> 	// TODO(andyk): Why does execution engine need mutation columns for Insert?
> 	fetchScope := mb.b.buildScan(
> ```

why not just use `mb.fetchScope` directly here instead of creating a new variable?
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 341 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MN--nvH8t4J0OGzAYQt:-MN--nvH8t4J0OGzAYQu:b-v2bhwo) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/optbuilder/mutation_builder.go#L341)):*
> ```Go
> 	}
> 
> 	// Set the fetchScope to the scope of the fetch expression.
> ```

[nit] what's a fetch expression?
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 408 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MN--qai7OPzPF4ueXTk:-MN--qai7OPzPF4ueXTl:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/optbuilder/mutation_builder.go#L408)):*
> ```Go
> 	mb.setFetchColIDs(mb.outScope.cols)
> 
> 	// Set the fetchScope to the scope of the fetch expression.
> ```

ditto
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 822 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MN--V7rCD4KroeLXDTD:-MN--V7rCD4KroeLXDTE:b-tvuy96) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/optbuilder/mutation_builder.go#L822)):*
> ```Go
> 			expr := mb.parsePartialIndexPredicateExpr(i)
> 
> 			// Build a synthesized PUT columns.
> ```

[nit] a synthesized -> synthesized
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 832 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MN--Z_f8Iw20epK8pkZ:-MN--Z_f8Iw20epK8pk_:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/optbuilder/mutation_builder.go#L832)):*
> ```Go
> 			}
> 
> 			// Build a synthesized DEL columns.
> ```

ditto


<!-- Sent from Reviewable.io -->

mgartner committed:
opt: fix FK cascades to child tables with partial indexes

Previously, the optimizer was not synthesizing partial index DEL columns
for FK cascading updates and deletes. As a result, a cascading `UPDATE`
could corrupt a child table's partial index, ultimately resulting in
incorrect query results. A cascading `DELETE` would not corrupt
partial indexes, but unnecessary `DEL` operations would be issued on
the partial index.

The optbuilder has been refactored so that these columns are correctly
projected. There are now three functions for projecting PUT columns, DEL
columns, and both PUT and DEL columns, each ensuring that the input
scopes are non-nil. These three functions are called from principal
functions in the optbuilder where CHECK constraint columns are also
projected, like `mutationBuilder.buildUpdate`. In theory this should
make it harder in the future to omit these necessary projections.

Additionally, the execution engine was unable to handle extraneous
columns that can be added as input to FK cascading updates. These
extraneous columns would be incorrectly interpreted as synthesized
partial index columns. This commit works around this issue by slicing
the source values with an upper bound in `updateNode.processSourceRow`.
The longer term fix is to not produce these columns (see issue #57097).

Fixes #57085
Fixes #57084

Release justification: This is a critical bug fix to a new feature,
partial indexes.

Release note (bug fix): A bug has been fixed that caused errors or
corrupted partial indexes of child tables in foreign key relationships
with cascading `UPDATE`s and `DELETE`s. The corrupt partial indexes
could result in incorrect query results. Any partial indexes on child
tables of foreign key relationships with `ON DELETE CASCADE` or `ON
UPDATE CASCADE` actions may be corrupt and should be dropped and
re-created. This bug was introduce in version 20.2.

mgartner reviewed (commented):
Thanks for the feedback @rytaft. I reworked this quite a bit, so some of them were no longer applicable. I think this is more polished now.

The first commit is from #57153.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-:-MN1GmcFEgrLfuCVyMFp:b-vdqmpi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/update.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MN-2_hnCMDTqeXNDcc_:-MN175WM9ORHpwgJRAIH:b2ku57d) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/update.go#L59)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is this getting used other than for its length?
</blockquote></details>

No. I realized this wasn't necessary and I've removed it.
___
*[pkg/sql/logictest/testdata/logic_test/partial_index, line 1530 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MN-1FmX9WQ9-vr_WAGX:-MN17vtr4C1Q9Sz3W9uA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/logictest/testdata/logic_test/partial_index#L1530)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should there be a similar test to ensure that rows are deleted from the partial index if group_id is updated from 3->4?
</blockquote></details>

Done.
___
*[pkg/sql/opt/exec/factory.opt, line 416 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MMzyRqe7LBJ7cIOng1T:-MN17xu34yyhTiPpwRiu:b-xiy3n1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/exec/factory.opt#L416)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] update comment to mention partial indexes
</blockquote></details>

I've removed this new argument — it wasn't necessary.
___
*[pkg/sql/opt/exec/execbuilder/mutation.go, line 343 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MMzybb30FdRXmvcT-Mx:-MN18CnN2xDlmHqFfi0w:b-3q7fij) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/exec/execbuilder/mutation.go#L343)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

why do we only need `Put` here and not `Del`?
</blockquote></details>

This has been removed — it wasn't necessary.
___
*[pkg/sql/opt/optbuilder/fk_cascade.go, line 637 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MMzz1fz77_7XZd5whcQ:-MN18J9P9PFL3vJwjXfo:b-2rqglh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/optbuilder/fk_cascade.go#L637)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

what's a fetch expression?
</blockquote></details>

I've updated the comment for the `fetchScope` field, and redone a lot of this code. LMK if it's still unclear what `fetchScope` is.
___
*[pkg/sql/opt/optbuilder/insert.go, line 908 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MMzzvrJ8LrH4k2o40bI:-MN18VNUArvszg1fJgWC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/optbuilder/insert.go#L908)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

why not just use `mb.fetchScope` directly here instead of creating a new variable?
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 341 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MN--nvH8t4J0OGzAYQt:-MN18Ys347kkCejs_su1:bu6xv6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/optbuilder/mutation_builder.go#L341)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] what's a fetch expression?
</blockquote></details>

See previous comment.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 408 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MN--qai7OPzPF4ueXTk:-MN18cYED0zxLAY6ESD6:bckp57o) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/optbuilder/mutation_builder.go#L408)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

See previous comment.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 822 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MN--V7rCD4KroeLXDTD:-MN18ksg0UtHItWP3kbB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/optbuilder/mutation_builder.go#L822)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] a synthesized -> synthesized
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 832 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MN--Z_f8Iw20epK8pkZ:-MN18lotA4HIjZ435j85:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1fef31e010d6f17cee7a8ea394f77bc22e0e630/pkg/sql/opt/optbuilder/mutation_builder.go#L832)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 6 of 15 files at r2, 18 of 18 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-:-MNOpCkt-j-s_mTjGYyX:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNOgh7p4FBVbP_SD7Lv:-MNOgh7q4WsCb9devCOC:b-7uu653) ([raw file](https://github.com/cockroachdb/cockroach/blob/99dd2be1656fc586daad5d62ee31c71c51c00b1b/pkg/sql/opt/optbuilder/mutation_builder.go#L797)):*
> ```Go
> 
> // partialIndexProjectColumns specifies which types of partial index columns to
> // project, PUT columns, DEL columns, or both. Different mutations require
> ```

[nit] project, -> project:
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 798 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNOgnO029qMdOqHSbxx:-MNOgnO11ivXeatvlC5R:baxkg3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/99dd2be1656fc586daad5d62ee31c71c51c00b1b/pkg/sql/opt/optbuilder/mutation_builder.go#L798)):*
> ```Go
> // partialIndexProjectColumns specifies which types of partial index columns to
> // project, PUT columns, DEL columns, or both. Different mutations require
> // different sets of these columns to be project. For example, INSERT does not
> ```

[nit] to be project -> to be projected


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-:-MNPuQtb2HJKM9CvVD8W:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 806 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNPsVbtA5cEF762XRuf:-MNPsVbtA5cEF762XRug:b-nctmke) ([raw file](https://github.com/cockroachdb/cockroach/blob/99dd2be1656fc586daad5d62ee31c71c51c00b1b/pkg/sql/opt/optbuilder/mutation_builder.go#L806)):*
> ```Go
> 	putCols partialIndexProjectColumns = 1 << iota
> 	delCols
> 	putAndDelCols = putCols | delCols
> ```

[nit] ` putAndDelCols partialIndexProjectColumns = ..`
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 811 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNPsret1QFnqahBX_uD:-MNPsret1QFnqahBX_uE:b-1hsd0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/99dd2be1656fc586daad5d62ee31c71c51c00b1b/pkg/sql/opt/optbuilder/mutation_builder.go#L811)):*
> ```Go
> // HasColumns tests whether pc contains all the partial index project columns in
> // other.
> func (pc partialIndexProjectColumns) HasColumns(other partialIndexProjectColumns) bool {
> ```

[nit] maybe just `Has` is a better name
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 817 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNPtgu4E-rON_ZlJvPM:-MNPtgu4E-rON_ZlJvPN:b-3exu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/99dd2be1656fc586daad5d62ee31c71c51c00b1b/pkg/sql/opt/optbuilder/mutation_builder.go#L817)):*
> ```Go
> // projectPartialIndexCols builds a Project that synthesizes boolean output
> // columns for each partial index defined on the target table. The projectCols
> // argument specifies if PUT columns, DEL columns, or both PUT and DEL columns
> ```

[nit] could use a description of what the scopes are
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 820 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNPtjjKCIcKpz9Rp-xr:-MNPtjjKCIcKpz9Rp-xs:b-8l7wgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/99dd2be1656fc586daad5d62ee31c71c51c00b1b/pkg/sql/opt/optbuilder/mutation_builder.go#L820)):*
> ```Go
> // should be projected.
> func (mb *mutationBuilder) projectPartialIndexCols(
> 	putScope *scope, delScope *scope, projectCols partialIndexProjectColumns,
> ```

An alternative here would be to provide three variants (`projectPartialIndex[Put|Del|PutAndDel]Cols`) and rename this to `projectPartialIndexColsImpl`. The variants would check for `nil` scopes, and `Impl` could just check them against `nil` instead of using `partialIndexProjectColumns`.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57100)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 797 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNOgh7p4FBVbP_SD7Lv:-MNQmbzz6NmQGPP7Wzwt:bbhaxth) ([raw file](https://github.com/cockroachdb/cockroach/blob/99dd2be1656fc586daad5d62ee31c71c51c00b1b/pkg/sql/opt/optbuilder/mutation_builder.go#L797)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] project, -> project:
</blockquote></details>

Remove this type entirely.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 798 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNOgnO029qMdOqHSbxx:-MNQmfCr8xcQPJcj3Dhk:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/99dd2be1656fc586daad5d62ee31c71c51c00b1b/pkg/sql/opt/optbuilder/mutation_builder.go#L798)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] to be project -> to be projected
</blockquote></details>

Removed.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 806 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNPsVbtA5cEF762XRuf:-MNQmZxMCS_tYdTUCDCg:b-f02ykq) ([raw file](https://github.com/cockroachdb/cockroach/blob/99dd2be1656fc586daad5d62ee31c71c51c00b1b/pkg/sql/opt/optbuilder/mutation_builder.go#L806)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] ` putAndDelCols partialIndexProjectColumns = ..`
</blockquote></details>

Removed this const entirely.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 811 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNPsret1QFnqahBX_uD:-MNQmY496YR7EtIpHvjr:b-fjxyij) ([raw file](https://github.com/cockroachdb/cockroach/blob/99dd2be1656fc586daad5d62ee31c71c51c00b1b/pkg/sql/opt/optbuilder/mutation_builder.go#L811)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] maybe just `Has` is a better name
</blockquote></details>

Removed this function entirely.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 817 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNPtgu4E-rON_ZlJvPM:-MNQmWxT7-efBENBsqAy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99dd2be1656fc586daad5d62ee31c71c51c00b1b/pkg/sql/opt/optbuilder/mutation_builder.go#L817)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] could use a description of what the scopes are
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 820 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNPtjjKCIcKpz9Rp-xr:-MNQmjmN2Rl4d-yqkBdf:bf1u4il) ([raw file](https://github.com/cockroachdb/cockroach/blob/99dd2be1656fc586daad5d62ee31c71c51c00b1b/pkg/sql/opt/optbuilder/mutation_builder.go#L820)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

An alternative here would be to provide three variants (`projectPartialIndex[Put|Del|PutAndDel]Cols`) and rename this to `projectPartialIndexColsImpl`. The variants would check for `nil` scopes, and `Impl` could just check them against `nil` instead of using `partialIndexProjectColumns`.
</blockquote></details>

I love it. Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 13 of 18 files at r3, 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-:-MNUAobk1aw247BVDYah:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 811 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNUAezb5W1qeB1RRAiF:-MNUAezb5W1qeB1RRAiG:bwmx7lp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f06aeaa421c06dd8c5172633594a049e4c7d36e6/pkg/sql/opt/optbuilder/mutation_builder.go#L811)):*
> ```Go
> // projectPartialIndexDelCols builds a Project that synthesizes boolean PUT
> // columns for each partial index defined on the target table. See
> // partialIndexPutColIDs for more info on these columns.
> ```

[nit] DelColIDs
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 837 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNUAk5q7Mz-AHnyTkVr:-MNUAk5q7Mz-AHnyTkVs:b2nhcjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f06aeaa421c06dd8c5172633594a049e4c7d36e6/pkg/sql/opt/optbuilder/mutation_builder.go#L837)):*
> ```Go
> 	}
> 	if delScope == nil {
> 		panic(errors.AssertionFailedf("cannot project partial index PUT columns with nil scope"))
> ```

DEL


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57100)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 811 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNUAezb5W1qeB1RRAiF:-MNUN6cL0HaIzXdpKDbd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f06aeaa421c06dd8c5172633594a049e4c7d36e6/pkg/sql/opt/optbuilder/mutation_builder.go#L811)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] DelColIDs
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/mutation_builder.go, line 837 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57100#-MNUAk5q7Mz-AHnyTkVr:-MNUNBnIDg0JxwN6QyZp:b-tnq7w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f06aeaa421c06dd8c5172633594a049e4c7d36e6/pkg/sql/opt/optbuilder/mutation_builder.go#L837)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

DEL
</blockquote></details>

Done. Fixed the one above in `projectPartialIndexDelCols` too.


<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+
